### PR TITLE
SOLR-14409: Existing violations allow bypassing policy rules when add…

### PR DESCRIPTION
**Pull request by noblepaul**

### CODE REVIEW

**Comment by noblepaul** [comment]
@sigram 
shall I go ahead and merge this?

File: solr/solrj/src/test/org/apache/solr/client/solrj/cloud/autoscaling/TestPolicy2.java [hunk 1]
```diff
@@ -516,6 +517,16 @@ public void testInfiniteLoop() {
     System.out.println(suggestions);
   }
 
+  public void testAddTooManyPerPolicy() {
+    Map<String, Object> m = (Map<String, Object>) loadFromResource("testAddTooManyPerPolicy.json");
+    SolrCloudManager cloudManagerFromDiagnostics = createCloudManagerFromDiagnostics(m);
+    AutoScalingConfig autoScalingConfig = new AutoScalingConfig((Map<String, Object>) getObjectByPath(m, false, "diagnostics/config"));
+    SolrException exp =  expectThrows(SolrException.class, () -> PolicyHelper.getReplicaLocations("TooManyPerPolicy", autoScalingConfig, cloudManagerFromDiagnostics,
+            EMPTY_MAP, Collections.singletonList("shard1"), 1, 0, 0, null));
+    assertTrue(exp.getMessage().contains("No node can satisfy the rules"));
```
Comment (madrob) [inline solr/solrj/src/test/org/apache/solr/client/solrj/cloud/autoscaling/TestPolicy2.java:526]: `assertThat(exp.getMessage(), contains("..."))`

Comment (noblepaul) [reply solr/solrj/src/test/org/apache/solr/client/solrj/cloud/autoscaling/TestPolicy2.java:526]: what's the difference?

**Comment by sigram** [comment]
This change fixes the original issue, so it's +1 on the change.

However, please note that the tests contain many unchecked casts and these need to be fixed before committing, otherwise gradle build will fail.

File: solr/solrj/src/test/org/apache/solr/client/solrj/cloud/autoscaling/TestPolicy2.java (hunk 1 above)
Comment (madrob) [reply solr/solrj/src/test/org/apache/solr/client/solrj/cloud/autoscaling/TestPolicy2.java:527]: You get a better error message when the test fails.

**Pull request merged**